requests-for-comments/2022-01-27-future-of-person-model.md

```diff
@@ -0,0 +1,108 @@
+# Request for Comments: Future of person model
+
+- **Created:** 2022-01-27
+- **Owner:** @macobo, @marcushyett-ph 
+
+## Problem: Scaling queries horizontally
+
+In posthog, having insights load fast is key for a great experience. However for large projects tracking >10M active persons per month query performance degrades fast.
+
+This can't be solved by scaling horizontally, because a lot of our queries do aggregation on persons - e.g. filtering by user properties, counting daily active persons and calculating funnels. 
+
+To do these queries we need to JOIN `person` and `person_distinct_id` tables. This in clickhouse is really slow (can cause >30x slower queries) and requires these tables to be replicated on every node, setting an upper bound to scalability. Teams with >10M users often run into memory usage problems as well.
+
+This RFC proposes both a technical and product solution to this problem: Store `person_id` and `person_properties` on the main events table.
+
+However, doing so will require breaking changes to the product. These changes are likely to cause some issues for customers, but these have either a negligible impact (affecting around 0.03% of events) or can be mitigated through documentation and education.
+
+## Proposed solution
+
+From a technical standpoint: Store `person_id` and `person_properties` on the main events table.
+
+From a product standpoint storing person_id and person_properties on events requires making the following changes:
+1. After the first session we will no longer merge events prior to login with the main person’s events.
```

Comment:
Am I interpreting this right? Is this saying that after making this change we'll no longer be able to merge an anoynmous user's timeline into the identified users?

So for example if I come to posthog.com as an anonymous user, distinct ID A and signup as distinct ID kunal@posthog.com – distinct ID A's will always look like a separate person from kunal@posthog.com?

Reply:
It depends.

The section below under customer impact goes into more detail.

To elaborate in different words though: during initial session (when kunal@ hasn't ever been identified) it behaves as you expect. The difference comes in on subsequent sessions e.g. on a different device where events pre login would get associated to a different person.

Let me know if this or the other explanation below make sense!

Reply:
Thanks - let's go a little deeper:

In today's world...

I come on incognito window from my phone:
- I have distinct ID A assigned. 
- I do a bunch of things before logging in
- I login, events performed under Distinct ID A can now be found looking at the timeline for kunal@posthog.com


I then come on from an incognito window from my laptop:
- I have distinct ID B assigned
- I do a bunch of things prior to login
- I login, events performed under Distinct ID B can now be found at the timeline for kunal@

So now looking at the event timeline for kunal@ - I see all the events that were performed from both distinct IDs A and B. 

Which pieces of this would break (if any) in the proposed change?

Reply:
@kpthatsme it's also worth reading [this doc from our friends at another YC analytics company](https://help.amplitude.com/hc/en-us/articles/115003135607-Track-unique-users-in-Amplitude) - as far as we can tell their merging system has a similar behavior - but they might have done a better job at explaining it

Reply:
Could we have some way to "backtrack" to the anonymous events/session post login?

For example, by adding the property `anon_distinct_id: "anonid2"` to events `11` and `12` in the table above. 

This would allow us to link session recordings with the pre-login part, and potentially allow some flexibility in the future when someone asks for a funnel one part of it is logged out events, and the other part consists of logged in events.

Reply:
TLDR: what if we delayed ingestion storing the anonymous user events somewhere and ingesting after identify event with the logged_in person ID. Then we'd still get our perf and simplicity improvements for queries without losing pre-logged in events being tied to the same person.

Long version:
Assumptions: Pre-login we don't know the logged_in person_id, hence events will be sent with a new uuid

There are only 3 options (let me know if you can think of any more):
### 1. join persons 
This is what we do now & has all the perf problems for queries. If we are not querying over a single person_id for unique users queries results in us joining persons. Adding any properties etc results in the same need to join persons. I'd bucket any ideas with running queries across CH/postgres here too & I think they all have the same perf problem.

#### 2. update events in clickhouse 
Updating the person_id from the anonymous one to the logged_in person_id: We can't really mutate events in CH, what we could do is run an async migration on the shard (essentially copy all data to a new table modifying the distinct_id's accordingly & then rename tables like 0002 async migration). This is probably expensive in terms of resources even done for a single shard; that migration would put us temporarily in a state of delayed ingestion (while we copy the last part of the table and rename we'd be queueing up new events in kafka similar to 0002).
#### 3. delay ingestion until we know the logged_in person_id: 

There are two options here which might not be all that bad of an idea:
1. delay by using some intermediary storage for logged_out users: we could dump all anonymous events somewhere keyed by uuid and then consume them once we see the identify event (I was thinking kafka, but we can't get messages by key, maybe by some separate table in ClickHouse) and we we can write the events to ClickHouse with the logged_in person_id. I was thinking up to certain delay, e.g. 1 h or 1 day could be longer in theory too, but note that during that delay we wouldn't see those events in queries (which could be acceptable). I wonder if this would actually not be a terrible idea?
2. Re-ingest the events with the new person_id and mark the other uuid as a deleted person (do we filter out deleted people somehow in queries or in any case filtering out is potentially cheaper than joins). We can occasionally run an async migration that clears them out.

-----
 
Additionally there's a problem tied to person properties and how they would get updated, but we could just decide on a global semantics that person properties can only be used on logged_in users (in any case this is probably less important and the funnels working etc is more important).

Reply:
Assuming at first we don't implement a solution to properly associate green events ("green" from diagram) to the proper user, is there a way we can mitigate impact to DAUs/WAUs/MAUs analytics? My biggest concern is that your active users metric becomes inaccurate because of this.

Reply:
> mitigate impact to DAUs/WAUs/MAUs analytics

perhaps filtering on logged in users would help?